core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java

```diff
@@ -93,7 +94,34 @@ public FillRangeRecordCursorFactory(
         this.timestampIndex = timestampIndex;
         this.valueFuncs = fillValues;
         this.metadata = metadata;
+
+        // only do this for value fill
+        if (!(valueFuncs.size() == 1 && valueFuncs.get(0).isNullConstant())) {
+            if (metadata.getColumnCount() - 1 > valueFuncs.size()) {
+                throw SqlException.$(0, "not enough fill values");
+            }
+
+            int passedTimestamp = 0;
+
+            // validate metadata
+            for (int i = 0; i < metadata.getColumnCount(); i++) {
+                int columnType = metadata.getColumnType(i);
+                if (i == metadata.getTimestampIndex() ||
+                        (columnType == ColumnType.TIMESTAMP && metadata.getTimestampIndex() == -1)) {
+                    passedTimestamp = 1;
+                    continue;
+                }
+
+                int fillColumnType = valueFuncs.get(i - passedTimestamp).getType();
+                if (fillColumnType != columnType && !ColumnType.isBuiltInWideningCast(fillColumnType, columnType)) {
+                    throw SqlException.$(0, "invalid fill value, cannot cast " + ColumnType.nameOf(fillColumnType) + " to " + ColumnType.nameOf(columnType));
```

Comment:
SqlException has a machanism to avoid string concatentation, please use it. cc @amunra